File: src/efficio_pptx_contracts/template_contract_migrations.py
"""Pure template contract migration planning over opaque tagged targets."""

from __future__ import annotations

import re
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from types import MappingProxyType

from ._template_contract_migration_catalog import (
    CURRENT_TEMPLATE_CONTRACT_REVISION,
    TEMPLATE_CONTRACT_REVISION_TAG,
    UNVERSIONED_TEMPLATE_CONTRACT_REVISION,
    RenameTagOperation,
    SetTagIfMissingOperation,
    TemplateContractMigration,
    TemplateContractMigrationCatalog,
    TemplateContractMigrationOperation,
    TemplateMigrationOperationType,
    TemplateTagScope,
    get_template_contract_migration_path,
    load_template_contract_migration_catalog,
)
from .errors import TemplateContractMigrationError

_REVISION = re.compile(r"^[0-9]+$")


@dataclass(frozen=True, slots=True)
class TemplateTagTarget:
    target_ref: str
    scope: TemplateTagScope
    tags: Mapping[str, str]


@dataclass(frozen=True, slots=True)
class TemplateTagPatch:
    target_ref: str
    scope: TemplateTagScope
    set_tags: Mapping[str, str]
    remove_tags: tuple[str, ...]


@dataclass(frozen=True, slots=True)
class TemplateContractMigrationPlan:
    from_revision: int
    to_revision: int
    patches: tuple[TemplateTagPatch, ...]


def plan_template_contract_migration(
    targets: Sequence[TemplateTagTarget],
) -> TemplateContractMigrationPlan:
    """Plan deterministic set/remove patches without mutating caller-owned tags."""
    prepared = _prepare_targets(targets)
    deck_indexes = [
        index for index, (target, _tags) in enumerate(prepared)
        if target.scope is TemplateTagScope.DECK
    ]
    if len(deck_indexes) != 1:
        raise TemplateContractMigrationError(
            "template migration requires exactly one deck target"
        )
    deck_index = deck_indexes[0]
    from_revision = _read_revision(prepared[deck_index][1])
    path = get_template_contract_migration_path(from_revision)
    original = [dict(tags) for _target, tags in prepared]
    for migration in path:
        for operation in migration.operations:
            _apply_operation(prepared, operation)
    prepared[deck_index][1][TEMPLATE_CONTRACT_REVISION_TAG] = str(
        CURRENT_TEMPLATE_CONTRACT_REVISION
    )
    _validate_no_retired_tags(prepared)
    patches = tuple(
        patch for index, (target, tags) in enumerate(prepared)
        if (patch := _build_patch(target, original[index], tags)) is not None
    )
    return TemplateContractMigrationPlan(
        from_revision=from_revision,
        to_revision=CURRENT_TEMPLATE_CONTRACT_REVISION,
        patches=patches,
    )


def _validate_no_retired_tags(
    targets: Sequence[tuple[TemplateTagTarget, Mapping[str, str]]],
) -> None:
    retired = {
        (operation.scope, operation.source_tag)
        for migration in load_template_contract_migration_catalog().migrations
        for operation in migration.operations
        if isinstance(operation, RenameTagOperation)
    }
    for target, tags in targets:
        if any(scope is target.scope and tag in tags for scope, tag in retired):
            raise TemplateContractMigrationError(
                "current template contract contains a retired tag"
            )


def _prepare_targets(
    targets: Sequence[TemplateTagTarget],
) -> list[tuple[TemplateTagTarget, dict[str, str]]]:
    prepared: list[tuple[TemplateTagTarget, dict[str, str]]] = []
    identities: set[tuple[TemplateTagScope, str]] = set()
    for target in targets:
        try:
            scope = TemplateTagScope(target.scope)
        except ValueError as error:
            raise TemplateContractMigrationError("template target scope is invalid") from error
        if not isinstance(target.target_ref, str) or not target.target_ref.strip():
            raise TemplateContractMigrationError("template target_ref must be non-empty")
        identity = (scope, target.target_ref)
        if identity in identities:
            raise TemplateContractMigrationError("template target references must be unique")
        identities.add(identity)
        if not isinstance(target.tags, Mapping) or any(
            not isinstance(key, str) or not isinstance(value, str)
            for key, value in target.tags.items()
        ):
            raise TemplateContractMigrationError("template target tags must map strings to strings")
        prepared.append((TemplateTagTarget(target.target_ref, scope, target.tags), dict(target.tags)))
    return prepared


def _read_revision(deck_tags: Mapping[str, str]) -> int:
    raw = deck_tags.get(TEMPLATE_CONTRACT_REVISION_TAG)
    if raw is None:
        return UNVERSIONED_TEMPLATE_CONTRACT_REVISION
    if _REVISION.fullmatch(raw) is None:
        raise TemplateContractMigrationError(
            f"{TEMPLATE_CONTRACT_REVISION_TAG} must be a non-negative integer string"
        )
    revision = int(raw)
    if revision > CURRENT_TEMPLATE_CONTRACT_REVISION:
        raise TemplateContractMigrationError(
            f"template contract revision {revision} is newer than supported revision "
            f"{CURRENT_TEMPLATE_CONTRACT_REVISION}"
        )
    return revision


def _apply_operation(
    targets: list[tuple[TemplateTagTarget, dict[str, str]]],
    operation: TemplateContractMigrationOperation,
) -> None:
    for target, tags in targets:
        if target.scope is not operation.scope:
            continue
        if isinstance(operation, SetTagIfMissingOperation):
            if operation.tag not in tags:
                tags[operation.tag] = operation.value
            continue
        source = tags.get(operation.source_tag)
        if source is None:
            continue
        mapped = source
        if operation.value_map is not None:
            mapped_value = operation.value_map.get(source)
            if mapped_value is None:
                raise TemplateContractMigrationError(
                    f"tag {operation.source_tag} has no migration mapping"
                )
            mapped = mapped_value
        existing = tags.get(operation.target_tag)
        if existing is not None and existing != mapped:
            raise TemplateContractMigrationError(
                f"tag {operation.target_tag} conflicts with migrated content"
            )
        tags[operation.target_tag] = mapped
        del tags[operation.source_tag]


def _build_patch(
    target: TemplateTagTarget,
    original: Mapping[str, str],
    migrated: Mapping[str, str],
) -> TemplateTagPatch | None:
    set_tags = {
        key: migrated[key]
        for key in sorted(migrated)
        if original.get(key) != migrated[key]
    }
    remove_tags = tuple(sorted(set(original) - set(migrated)))
    if not set_tags and not remove_tags:
        return None
    return TemplateTagPatch(
        target_ref=target.target_ref,
        scope=target.scope,
        set_tags=MappingProxyType(set_tags),
        remove_tags=remove_tags,
    )


__all__ = [
    "CURRENT_TEMPLATE_CONTRACT_REVISION", "TEMPLATE_CONTRACT_REVISION_TAG",
    "UNVERSIONED_TEMPLATE_CONTRACT_REVISION", "RenameTagOperation",
    "SetTagIfMissingOperation", "TemplateContractMigration",
    "TemplateContractMigrationCatalog", "TemplateContractMigrationOperation",
    "TemplateContractMigrationPlan", "TemplateContractMigrationError",
    "TemplateMigrationOperationType", "TemplateTagPatch", "TemplateTagScope",
    "TemplateTagTarget", "get_template_contract_migration_path",
    "load_template_contract_migration_catalog", "plan_template_contract_migration",
]

File: ts/editor/template-contract-migrations.ts
import { templateContractMigrationCatalog } from "../../generated/ts/presentation/templateContractMigrations.js";
import { copyContractValue } from "./contract-copy.js";

export const TEMPLATE_CONTRACT_REVISION_TAG = "efficio_template_contract_revision";
export const UNVERSIONED_TEMPLATE_CONTRACT_REVISION = 0;
export const CURRENT_TEMPLATE_CONTRACT_REVISION = templateContractMigrationCatalog.current_revision;

export type TemplateTagScope = "deck" | "slide" | "shape";
export type SetTagIfMissingOperation = Readonly<{
  type: "set_tag_if_missing";
  scope: TemplateTagScope;
  tag: string;
  value: string;
}>;
export type RenameTagOperation = Readonly<{
  type: "rename_tag";
  scope: TemplateTagScope;
  source_tag: string;
  target_tag: string;
  value_map?: Readonly<Record<string, string>>;
}>;
export type TemplateContractMigrationOperation = SetTagIfMissingOperation | RenameTagOperation;
export type TemplateContractMigration = Readonly<{
  format_version: 1;
  contract_type: "template_contract_migration";
  from_revision: number;
  to_revision: number;
  description: string;
  operations: readonly TemplateContractMigrationOperation[];
}>;
export type TemplateContractMigrationCatalog = Readonly<{
  generated_from: readonly string[];
  contract_type: "template_contract_migrations";
  format_version: 1;
  unversioned_revision: 0;
  current_revision: number;
  revision_tag: typeof TEMPLATE_CONTRACT_REVISION_TAG;
  migrations: readonly TemplateContractMigration[];
}>;
export type TemplateTagTarget = Readonly<{
  target_ref: string;
  scope: TemplateTagScope;
  tags: Readonly<Record<string, string>>;
}>;
export type TemplateTagPatch = Readonly<{
  target_ref: string;
  scope: TemplateTagScope;
  set_tags: Readonly<Record<string, string>>;
  remove_tags: readonly string[];
}>;
export type TemplateContractMigrationPlan = Readonly<{
  from_revision: number;
  to_revision: number;
  patches: readonly TemplateTagPatch[];
}>;

export class TemplateContractMigrationError extends Error {
  override readonly name = "TemplateContractMigrationError";
}

const catalog = templateContractMigrationCatalog as unknown as TemplateContractMigrationCatalog;

export function getTemplateContractMigrationCatalog(): TemplateContractMigrationCatalog {
  return copyContractValue(catalog);
}

export function getTemplateContractMigrationPath(
  fromRevision: number,
): readonly TemplateContractMigration[] {
  validateRevision(fromRevision);
  const bySource = new Map(catalog.migrations.map((migration) => [migration.from_revision, migration]));
  const path: TemplateContractMigration[] = [];
  let revision = fromRevision;
  while (revision < CURRENT_TEMPLATE_CONTRACT_REVISION) {
    const migration = bySource.get(revision);
    if (!migration) {
      throw new TemplateContractMigrationError(
        `No template contract migration exists from revision ${revision}.`,
      );
    }
    path.push(copyContractValue(migration));
    revision = migration.to_revision;
  }
  return path;
}

export function planTemplateContractMigration(
  targets: readonly TemplateTagTarget[],
): TemplateContractMigrationPlan {
  const prepared = prepareTargets(targets);
  const deckIndexes = prepared.flatMap((entry, index) => entry.target.scope === "deck" ? [index] : []);
  if (deckIndexes.length !== 1) {
    throw new TemplateContractMigrationError("Template migration requires exactly one deck target.");
  }
  const deckIndex = deckIndexes[0];
  const fromRevision = readRevision(prepared[deckIndex].tags);
  const path = getTemplateContractMigrationPath(fromRevision);
  const originals = prepared.map(({ tags }) => ({ ...tags }));
  for (const migration of path) {
    for (const operation of migration.operations) applyOperation(prepared, operation);
  }
  prepared[deckIndex].tags[TEMPLATE_CONTRACT_REVISION_TAG] = String(CURRENT_TEMPLATE_CONTRACT_REVISION);
  validateNoRetiredTags(prepared);
  const patches = prepared.flatMap(({ target, tags }, index) => {
    const patch = buildPatch(target, originals[index], tags);
    return patch ? [patch] : [];
  });
  return { from_revision: fromRevision, to_revision: CURRENT_TEMPLATE_CONTRACT_REVISION, patches };
}

function validateNoRetiredTags(targets: PreparedTarget[]): void {
  for (const migration of catalog.migrations) {
    for (const operation of migration.operations) {
      if (operation.type !== "rename_tag") continue;
      if (targets.some(({ target, tags }) =>
        target.scope === operation.scope && operation.source_tag in tags
      )) {
        throw new TemplateContractMigrationError(
          "Current template contract contains a retired tag.",
        );
      }
    }
  }
}

type PreparedTarget = { target: TemplateTagTarget; tags: Record<string, string> };

function prepareTargets(targets: readonly TemplateTagTarget[]): PreparedTarget[] {
  if (!Array.isArray(targets)) {
    throw new TemplateContractMigrationError("Template targets must be an array.");
  }
  const identities = new Set<string>();
  return targets.map((target) => {
    if (typeof target !== "object" || target === null) {
      throw new TemplateContractMigrationError("Template target must be an object.");
    }
    if (typeof target.target_ref !== "string" || !target.target_ref.trim()) {
      throw new TemplateContractMigrationError("Template target_ref must be non-empty.");
    }
    if (!(["deck", "slide", "shape"] as const).includes(target.scope)) {
      throw new TemplateContractMigrationError("Template target scope is invalid.");
    }
    const identity = `${target.scope}\u0000${target.target_ref}`;
    if (identities.has(identity)) {
      throw new TemplateContractMigrationError("Template target references must be unique.");
    }
    identities.add(identity);
    if (typeof target.tags !== "object" || target.tags === null || Array.isArray(target.tags)) {
      throw new TemplateContractMigrationError("Template target tags must map strings to strings.");
    }
    if (Object.values(target.tags).some((value) => typeof value !== "string")) {
      throw new TemplateContractMigrationError("Template target tags must map strings to strings.");
    }
    return { target: { ...target, tags: { ...target.tags } }, tags: { ...target.tags } };
  });
}

function readRevision(deckTags: Readonly<Record<string, string>>): number {
  const raw = deckTags[TEMPLATE_CONTRACT_REVISION_TAG];
  if (raw === undefined) return UNVERSIONED_TEMPLATE_CONTRACT_REVISION;
  if (!/^[0-9]+$/.test(raw)) {
    throw new TemplateContractMigrationError(
      `${TEMPLATE_CONTRACT_REVISION_TAG} must be a non-negative integer string.`,
    );
  }
  const revision = Number(raw);
  validateRevision(revision);
  return revision;
}

function validateRevision(revision: number): void {
  if (!Number.isSafeInteger(revision) || revision < 0) {
    throw new TemplateContractMigrationError("Template contract revision must be non-negative.");
  }
  if (revision > CURRENT_TEMPLATE_CONTRACT_REVISION) {
    throw new TemplateContractMigrationError(
      `Template contract revision ${revision} is newer than supported revision ${CURRENT_TEMPLATE_CONTRACT_REVISION}.`,
    );
  }
}

function applyOperation(targets: PreparedTarget[], operation: TemplateContractMigrationOperation): void {
  for (const { target, tags } of targets) {
    if (target.scope !== operation.scope) continue;
    if (operation.type === "set_tag_if_missing") {
      if (!(operation.tag in tags)) tags[operation.tag] = operation.value;
      continue;
    }
    const source = tags[operation.source_tag];
    if (source === undefined) continue;
    const mapped = operation.value_map ? operation.value_map[source] : source;
    if (mapped === undefined) {
      throw new TemplateContractMigrationError(`Tag ${operation.source_tag} has no migration mapping.`);
    }
    const existing = tags[operation.target_tag];
    if (existing !== undefined && existing !== mapped) {
      throw new TemplateContractMigrationError(
        `Tag ${operation.target_tag} conflicts with migrated content.`,
      );
    }
    tags[operation.target_tag] = mapped;
    delete tags[operation.source_tag];
  }
}

function buildPatch(
  target: TemplateTagTarget,
  original: Readonly<Record<string, string>>,
  migrated: Readonly<Record<string, string>>,
): TemplateTagPatch | undefined {
  const set_tags = Object.fromEntries(
    Object.keys(migrated).sort().flatMap((key) => original[key] === migrated[key] ? [] : [[key, migrated[key]]]),
  );
  const remove_tags = Object.keys(original).filter((key) => !(key in migrated)).sort();
  if (Object.keys(set_tags).length === 0 && remove_tags.length === 0) return undefined;
  return { target_ref: target.target_ref, scope: target.scope, set_tags, remove_tags };
}
